com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```diff
@@ -0,0 +1,769 @@
+    // Therefore, the following revision will either be the revision when item was deleted or the
+    // the revision when item was lastly changed while in the recycle bin.
+    item.setRevision(workItem.getLastRevision());
+    return item;
+  }
+
+  /** Main method that will process the work item history based on the parameters in the request */
+  private WorkItemForJson processHistory(IWorkItem workItem, IDataService dataService) {
+
+    // Short circuit for performance reasons, don't need to make Polarion fetch history
+    if (Integer.valueOf(workItem.getLastRevision()) <= Integer.valueOf(lastUpdate)) {
+      return null;
+    }
+
+    WorkItemForJson workItemForJson = null;
+
+    IPObjectList<IWorkItem> workItemHistory = dataService.getObjectHistory(workItem);
+    if (workItemHistory != null) {
+      if (workItemHistory.size() == 1 && workItemHistory.get(0) != null) {
+        // No changes in history when size == 1 (the WI remains as created)
+        // We then return only if the item was created within the revision boundaries of the request
+        if (Integer.valueOf(workItemHistory.get(0).getRevision()) <= Integer.valueOf(endRevision))
+          workItemForJson =
+              Utils.castWorkItem(workItemHistory.get(0), includeCustomFields, includeLinkRoles);
+      } else if (workItemHistory.size() > 1) {
+        /**
+         * From Polarion JavaDoc: "The history list is sorted from the oldest (first) to the newest
+         * (last)."
+         * https://almdemo.polarion.com/polarion/sdk/doc/javadoc/com/polarion/platform/persistence/IDataService.html#getObjectHistory(T)
+         * Then, we get the last one from the history as the current revision
+         */
+        int lastUpdateIndex = searchIndexWorkItemHistory(workItemHistory);
+        IDiffManager diffManager = dataService.getDiffManager();
+        Collection<WorkItemChange> workItemChanges = new ArrayList<WorkItemChange>();
+        int endIndex =
+            collectWorkItemChanges(
+                workItemChanges, workItem.getId(), workItemHistory, diffManager, lastUpdateIndex);
+        // Using the endIndex to return the workItem as in the endRevision # (not necessarily the
+        // latest version of the item)
+        if (endIndex >= 0) {
+          workItemForJson =
+              Utils.castWorkItem(
+                  workItemHistory.get(endIndex), includeCustomFields, includeLinkRoles);
+          workItemForJson.setWorkItemChanges(workItemChanges);
+        }
+      } else {
+        /**
+         * No history. Empty list. From Polarion JavaDoc: "An empty list is returned if the object
+         * does not support history retrieval."
+         * "https://almdemo.polarion.com/polarion/sdk/doc/javadoc/com/polarion/platform/persistence/IDataService.html#getObjectHistory(T)"
+         */
+        // Since we're grabbing WIs (and they support history retrieval), as far as we can tell from
+        // Polarion docs, this else will never execute.
```

Comment:
Yeah, if that ever happens it's a Polarion server failure. So, I raised an exception and on the outermost method I catch, log, and return a 500 error to the client. The rationale is: if we cannot return the history of an item (which is that else), it's better to not fulfill the request and return a server error rather than skip that workr item and return a state that does not necessarily reflect the work item history and potentially lead to inconsistencies on the client side.

Done.